dependency-analysis/main.go

```diff
@@ -0,0 +1,349 @@
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-github/v46/github"
+	"golang.org/x/oauth2"
+)
+
+func main() {
+	vulnerabilities := ""
+	result := ""
+
+	repo := os.Getenv("GITHUB_REPOSITORY")
+	commitSHA := os.Getenv("GITHUB_SHA")
+	token := os.Getenv("GITHUB_TOKEN")
+	pr := os.Getenv("GITHUB_PR_NUMBER")
+	ghUser := os.Getenv("GITHUB_ACTOR")
+	if err := Validate(token, repo, commitSHA, pr); err != nil {
+		log.Fatal(err)
+	}
+
+	ownerRepo := strings.Split(repo, "/")
+	owner := ownerRepo[0]
+	repo = ownerRepo[1]
+	checks, err := GetScorecardChecks()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defaultBranch, err := getDefaultBranch(owner, repo, token)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: token})
+	tc := oauth2.NewClient(context.Background(), ts)
+	client := github.NewClient(tc)
+	data, err := GetDependencyDiff(owner, repo, token, defaultBranch, commitSHA)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	m := make(map[string]DependencyDiff)
+	for _, dep := range *data { //nolint:gocritic
+		m[dep.SourceRepositoryURL] = dep
+	}
+
+	for k, i := range m { //nolint:gocritic
+		url := strings.TrimPrefix(k, "https://")
+		scorecard, err := GetScore(url)
+		if err != nil && len(i.Vulnerabilities) > 0 {
+			sb := strings.Builder{}
+			sb.WriteString(fmt.Sprintf("<details><summary>Vulnerabilties %s</summary>\n </br>", i.SourceRepositoryURL))
+			sb.WriteString("<table>\n")
+			sb.WriteString("<tr>\n")
+			sb.WriteString("<th>Severity</th>\n")
+			sb.WriteString("<th>AdvisoryGHSAId</th>\n")
+			sb.WriteString("<th>AdvisorySummary</th>\n")
+			sb.WriteString("<th>AdvisoryUrl</th>\n")
+			sb.WriteString("</tr>\n")
+			for _, v := range i.Vulnerabilities {
+				sb.WriteString("<tr>\n")
+				sb.WriteString(fmt.Sprintf("<td>%s</td>\n", v.Severity))
+				sb.WriteString(fmt.Sprintf("<td>%s</td>\n", v.AdvisoryGHSAId))
+				sb.WriteString(fmt.Sprintf("<td>%s</td>\n", v.AdvisorySummary))
+				sb.WriteString(fmt.Sprintf("<td>%s</td>\n", v.AdvisoryURL))
+			}
+			sb.WriteString("</table>\n")
+			sb.WriteString("</details>\n")
+			vulnerabilities += sb.String()
+			continue
+		}
+		scorecard.Checks = filter(scorecard.Checks, func(check Check) bool {
+			for _, c := range checks {
+				if check.Name == c {
+					return true
+				}
+			}
+			return false
+		})
+		scorecard.Vulnerabilities = i.Vulnerabilities
+		result += GitHubIssueComment(&scorecard)
+	}
+	// convert pr to int
+	prInt, err := strconv.Atoi(pr)
+	if err != nil {
+		log.Fatal(err)
+	}
+	// create or update comment
+	if vulnerabilities == "" && result == "" {
+		return
+	}
+	if err := createOrUpdateComment(
+		client, owner, ghUser, repo, prInt, "## Scorecard Results</br>\n"+vulnerabilities+"</br>"+result); err != nil {
```

Comment:
ghUser is `GITHUB_ACTOR` which is the user who triggered the workflow - what do we expect GITHUB_ACTOR to be usually? if the workflow is re-triggered manually, ghUser would be the identity of a human repo maintainer - and then we'd try to update their first comment, it seems?

Reply:
The`GITHUB_ACTOR` https://docs.github.com/en/actions/learn-github-actions/variables would be different if a human runs it, so we won't be updating the existing comment but will be creating new one.

Do you have other thoughts?

Reply:
If we use a hardcoded username (like "scorecard-dependencydiff" for example) instead of `GITHUB_ACTOR`, it would avoid the issue where DependencyDiff attempts to post a comment with the name of the user who triggered the action.

Reply:
My logic searches for GitHub comments by the USER, and AFAIK, I cannot randomly assign a user that would create a comment with that user as a reference. So the comment would be created by the `GITHUB_ACTOR`. Unless you think it is possible?

Reply:
Ah I just found the docs that mention the difference between `actor` (which stays the same when your workflow is rerun manually) and `triggering_actor`. This is fine.